src/c_arc.rs

```diff
@@ -0,0 +1,168 @@
+            let len = mem::size_of_val(real_ref);
+            let out_of_bounds = offset > len;
+            if out_of_bounds {
+                dbg!(real_ref as *const T);
+                dbg!(self.stable_ref.as_ptr());
+                dbg!(offset);
+                dbg!(len);
+                panic!("CArc::stable_ref is out of bounds");
+            }
+        }
+
+        // Safety: [`Self::stable_ref`] is a ptr
+        // derived from [`Self::owner`]'s through [`CBox::as_ref`]
+        // and is thus safe to dereference.
+        unsafe { self.stable_ref.as_ref() }
+    }
+}
+
+impl<T: ?Sized> Deref for CArc<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        self.as_ref()
+    }
+}
+
+impl<T: ?Sized> Clone for CArc<T> {
+    fn clone(&self) -> Self {
+        let Self {
+            owner,
+            #[cfg(debug_assertions)]
+            base_stable_ref,
+            stable_ref,
+        } = self;
+        Self {
+            owner: owner.clone(),
+            #[cfg(debug_assertions)]
+            base_stable_ref: base_stable_ref.clone(),
+            // Safety: The ref remains stable across an [`Arc::clone`].
+            stable_ref: stable_ref.clone(),
+        }
+    }
+}
+
+impl<T: ?Sized> From<Arc<CBox<T>>> for CArc<T> {
+    fn from(owner: Arc<CBox<T>>) -> Self {
+        let stable_ref = (*owner).as_ref().into();
+        Self {
+            owner,
+            #[cfg(debug_assertions)]
+            base_stable_ref: stable_ref,
+            stable_ref,
+        }
+    }
+}
+
+impl<T: ?Sized> CArc<T> {
+    pub fn wrap(owner: CBox<T>) -> Rav1dResult<Self> {
+        let owner = Arc::new(owner); // TODO fallible allocation
+        Ok(owner.into())
+    }
+}
+
+/// An opaque, raw [`CArc`], which is represented as a [`NonNull`] ptr to a [`CBox`].
+///
+/// To keep the type FFI-safe, a [`PhantomData`] wrapper is used,
+/// but the ptr is actually a [`CBox`].
+#[repr(transparent)]
+pub struct RawCArc<T: ?Sized>(NonNull<PhantomData<CBox<T>>>);
+
+impl<T: ?Sized> CArc<T> {
+    /// Convert into a raw, opaque form suitable for C FFI.
+    pub fn into_raw(self) -> RawCArc<T> {
+        let raw = Arc::into_raw(self.owner);
+        // Safety: [`Arc::into_raw`] guarantees it always returns non-null ptrs.
+        let raw = unsafe { NonNull::new_unchecked(raw.cast_mut()) };
+        RawCArc(raw.cast())
+    }
+
+    /// # Safety
+    ///
+    /// The [`RawCArc`] must be originally from [`Self::into_raw`].
+    #[deny(unsafe_op_in_unsafe_fn)]
+    pub unsafe fn from_raw(raw: RawCArc<T>) -> Self {
+        // Safety: The [`RawCArc`] contains the output of [`Arc::into_raw`],
+        // so we can call [`Arc::from_raw`] on it.
+        let owner = unsafe { Arc::from_raw(raw.0.cast::<CBox<T>>().as_ptr()) };
+        owner.into()
+    }
+}
+
+impl<T> AddAssign<usize> for CArc<[T]> {
+    /// Slice [`Self::stable_ref`].
+    fn add_assign(&mut self, rhs: usize) {
+        self.stable_ref = self[rhs..].into();
+    }
+}
```

Comment:
I think we should make this our own method, rather than using `+=`. There's not a good way to document this and the behavior wasn't obvious to me. This tripped me up, and we could make it more explicit what we are doing when we shift the start of the slice view.

Reply:
Okay I can do that.  I wanted to make it so we could slice it like a normal slice but that didn't work out with the `Deref` `trait`.

Reply:
I replaced it with `fn slice_in_place` in 6df10fa, which takes a normal slice range, so we'll be able to do things like `data.slice_in_place(offset..);` or `data.slice_in_place(..len)`.  It's more flexible in that it supports any kind of range.  This necessitates some extra `..`s sometimes, but I think it's worth it because it's similar syntactically to normal slicing.